crates/jstz_proto/src/api/smart_function.rs

```diff
@@ -87,27 +91,33 @@ impl SmartFunction {
         Ok(address.to_string())
     }
 
+    // Invariant: The function should always be called within a js_host_context
     fn call(
         self_address: &Address,
         request: &JsNativeObject<Request>,
         operation_hash: OperationHash,
         context: &mut Context<'_>,
     ) -> JsResult<JsValue> {
         // 1. Get address from request
-        let address = request
-            .deref()
-            .url()
-            .domain()
-            .and_then(|domain| Address::from_base58(domain).ok())
-            .ok_or_else(|| {
-                JsError::from_native(JsNativeError::error().with_message("Invalid host"))
-            })?;
-
-        // 2. Set the referer of the request to the current smart function address
-        headers::test_and_set_referrer(&request.deref(), self_address)?;
+        let mut request_deref = request.deref_mut();
+        let domain = request_deref.url().domain().ok_or_else(|| {
+            JsError::from_native(JsNativeError::error().with_message("Invalid host"))
+        })?;
+        match domain {
+            JSTZ_HOST => SystemScript::run(self_address, &mut request_deref, context),
```

Comment:
I like the name, but I find `SystemScript` inconsistent. Perhaps `JstzScript`, `HostScript` (akin to host functions), or `NativeScript` (akin to `NativeClass`, `NativeFunction`)?

Reply:
I've gone w `HostScript`